{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"extract_data.ipynb","provenance":[],"collapsed_sections":[]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"code","metadata":{"id":"PkvyfaVx0SsM","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":34},"outputId":"e2886e47-407a-45ea-9b03-5254766b1ca7","executionInfo":{"status":"ok","timestamp":1575401153228,"user_tz":300,"elapsed":428,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}}},"source":["\"\"\"Extract gene sequence, translate in alternate reading frames to get non-gene sequence\"\"\""],"execution_count":1,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'Extract gene sequence, translate in alternate reading frames to get non-gene sequence'"]},"metadata":{"tags":[]},"execution_count":1}]},{"cell_type":"code","metadata":{"id":"z7dlMqVT0Ewf","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":51},"outputId":"c2678e4b-9254-4442-ab3c-673ccc739edf","executionInfo":{"status":"ok","timestamp":1575401155896,"user_tz":300,"elapsed":3081,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}}},"source":["!pip install biopython"],"execution_count":2,"outputs":[{"output_type":"stream","text":["Requirement already satisfied: biopython in /usr/local/lib/python3.6/dist-packages (1.75)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (from biopython) (1.17.4)\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"H-hILFuvCoWL","colab_type":"code","colab":{}},"source":["import os\n","import re\n","import gzip\n","import pickle\n","from Bio import SeqIO\n","from Bio.Data import CodonTable"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"Rdk5h5g6CoRV","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":34},"outputId":"80395320-0587-463e-b490-b295915327ed","executionInfo":{"status":"ok","timestamp":1575401155898,"user_tz":300,"elapsed":3063,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}}},"source":["# Load the Drive helper and mount\n","from google.colab import drive\n","\n","# This will prompt for authorization.\n","drive.mount('/content/drive')"],"execution_count":4,"outputs":[{"output_type":"stream","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"Zd2eMhwd0Sn4","colab_type":"code","colab":{}},"source":["def check_known_protein(protein_description):\n","    \"\"\"filter out hypothetical and putative proteins\"\"\"\n","    is_hypothetical = ('hypoth' in protein_description or 'hetical' in protein_description) # deals with most common misspellings of \"hypothetical\"\n","    is_putative = 'putative' in protein_description\n","    is_known = (not is_hypothetical) and (not is_putative)\n","    return is_known\n","\n","def get_CDS_locations(genome_path):\n","    \"\"\"exctracts gene locations and seqs from known protein coding genes, only non-hypothetical\"\"\"\n","    gene_loc_list = []\n","    gene_seq_list = []\n","    with gzip.open(genome_path, 'rt') as handle: # open file\n","        for record in SeqIO.parse(handle, \"genbank\"): # parse each record within file\n","            feature_list = record.features\n","            CDS_list = [x for x in feature_list if x.type=='CDS'] # get all CDS features in record\n","            for CDS in CDS_list:\n","                try:\n","                    protein_description = CDS.qualifiers['product'][0]\n","                    if check_known_protein(protein_description): # only use non-hypothetical proteins\n","                        gene_loc = CDS.location\n","                        gene_loc_list.append(gene_loc)\n","\n","                        gene_seq = CDS.extract(record.seq) # extract locations of CDS\n","                        gene_seq_list.append(gene_seq)\n","                except:\n","                    pass\n","\n","    return gene_loc_list, gene_seq_list"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"bLbJdOKn0SlZ","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":51},"outputId":"ef7a7501-7480-4d19-a73a-37bba29c6ce9","executionInfo":{"status":"ok","timestamp":1575401157422,"user_tz":300,"elapsed":4549,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}}},"source":["# load fasta sequence\n","d = \"drive/My Drive/Colab Notebooks/smaug/data/ecoli_MG1655\"\n","genome_path = os.path.join(d, \"GCF_000005845.2_ASM584v2_genomic.fna.gz\")\n","\n","with gzip.open(genome_path, 'rt') as handle:\n","    for record in SeqIO.parse(handle, \"fasta\"):\n","        seq = record.seq\n","\n","# get gene locations\n","gbff_path = os.path.join(d, \"GCF_000005845.2_ASM584v2_genomic.gbff.gz\")\n","gene_loc_list, gene_seq_list = get_CDS_locations(gbff_path)\n","\n","# total genes and example loc\n","print(len(gene_loc_list))\n","print(gene_loc_list[300])"],"execution_count":6,"outputs":[{"output_type":"stream","text":["3532\n","[381350:381716](+)\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"gLIjqaFu0SgY","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":51},"outputId":"e126e9bf-087a-44ef-b184-14bf5b214618","executionInfo":{"status":"ok","timestamp":1575401157859,"user_tz":300,"elapsed":4240,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}}},"source":["# format genes correctly\n","gene_aa = [x.translate(table=11, to_stop=False) for x in gene_seq_list]\n","gene_aa_filtered = [x[1:-1][::-1] for x in gene_aa if (x[-1] == \"*\")] # cut off start and stop codons, and reverse (strand should be 3' to 5' for model)"],"execution_count":7,"outputs":[{"output_type":"stream","text":["/usr/local/lib/python3.6/dist-packages/Bio/Seq.py:2748: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n","  BiopythonWarning)\n"],"name":"stderr"}]},{"cell_type":"code","metadata":{"id":"ug_AI-9o0Se3","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":51},"outputId":"526abdb1-33d1-4386-a579-bc2c39ba378a","executionInfo":{"status":"ok","timestamp":1575401161493,"user_tz":300,"elapsed":7546,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}}},"source":["# get subORFs in all alternate reading frames\n","arf_1 = [x[1:].translate(table=11, to_stop=False) for x in gene_seq_list]\n","arf_2 = [x[2:].translate(table=11, to_stop=False) for x in gene_seq_list]\n","arf_r0 = [x.reverse_complement()[0:].translate(table=11, to_stop=False) for x in gene_seq_list]\n","arf_r1 = [x.reverse_complement()[1:].translate(table=11, to_stop=False) for x in gene_seq_list]\n","arf_r2 = [x.reverse_complement()[2:].translate(table=11, to_stop=False) for x in gene_seq_list]"],"execution_count":8,"outputs":[{"output_type":"stream","text":["/usr/local/lib/python3.6/dist-packages/Bio/Seq.py:2748: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n","  BiopythonWarning)\n"],"name":"stderr"}]},{"cell_type":"code","metadata":{"id":"BYvcVwm00Sbj","colab_type":"code","colab":{}},"source":["def extract_fake_ORFs(seq_aa):\n","    stop_positions = [x.start() for x in re.finditer(\"\\\\*\", str(seq_aa))]\n","    try:\n","        fake_ORFs = [str(seq_aa[stop_positions[i-1]+1 : x])[::-1] for i, x in enumerate(stop_positions)] # cut off stop codons, and reverse (strand should be 3' to 5' for model)\n","        fake_ORFs = [x for x in fake_ORFs if len(x) > 40] # enforce minimum ORF length\n","    except:\n","        return []\n","    return fake_ORFs\n","\n","fake_aa = []\n","# for i, arf in enumerate(arf_1):\n","#     print(i)\n","#     if i == 10: \n","#         break\n","arfake_1 = [extract_fake_ORFs(x) for x in arf_1]\n","arfake_2 = [extract_fake_ORFs(x) for x in arf_2]\n","arfake_r0 = [extract_fake_ORFs(x) for x in arf_r0]\n","arfake_r1 = [extract_fake_ORFs(x) for x in arf_r1]\n","arfake_r2 = [extract_fake_ORFs(x) for x in arf_r2]\n","\n","for x in arfake_1:\n","    fake_aa.extend(x)\n","for x in arfake_2:\n","    fake_aa.extend(x)\n","for x in arfake_r0:\n","    fake_aa.extend(x)\n","for x in arfake_r1:\n","    fake_aa.extend(x)\n","for x in arfake_r2:\n","    fake_aa.extend(x)"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"4RpNGPW40SWQ","colab_type":"code","colab":{}},"source":["# save\n","d = \"drive/My Drive/Colab Notebooks/smaug/data\"\n","gene_ORF_path = os.path.join(d, \"ecoli_MG1655_geneORFs.pkl\")\n","fake_ORF_path = os.path.join(d, \"ecoli_MG1655_fakeORFs.pkl\")\n","\n","with open(gene_ORF_path, 'wb') as f:\n","    pickle.dump(gene_aa_filtered, f)\n","with open(fake_ORF_path, 'wb') as f:\n","    pickle.dump(fake_aa, f)"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"o3dhPqbW0ST5","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"TCn8t_KeQTpc","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]}]}