{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"smaug.ipynb","provenance":[],"collapsed_sections":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"accelerator":"GPU"},"cells":[{"cell_type":"code","metadata":{"id":"w0uUIeponHgz","colab_type":"code","outputId":"184a7805-21e8-4ee1-f3fd-eeaa5634c3cc","executionInfo":{"status":"ok","timestamp":1574786443701,"user_tz":300,"elapsed":14825,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}},"colab":{"base_uri":"https://localhost:8080/","height":714}},"source":["!pip install biopython\n","!pip install transformers"],"execution_count":1,"outputs":[{"output_type":"stream","text":["Collecting biopython\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/96/01/7e5858a1e54bd0bd0d179cd74654740f07e86fb921a43dd20fb8beabe69d/biopython-1.75-cp36-cp36m-manylinux1_x86_64.whl (2.3MB)\n","\u001b[K     |████████████████████████████████| 2.3MB 2.8MB/s \n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (from biopython) (1.17.4)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.75\n","Collecting transformers\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/fd/f9/51824e40f0a23a49eab4fcaa45c1c797cbf9761adedd0b558dab7c958b34/transformers-2.1.1-py3-none-any.whl (311kB)\n","\u001b[K     |████████████████████████████████| 317kB 2.8MB/s \n","\u001b[?25hCollecting regex\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/e3/8e/cbf2295643d7265e7883326fb4654e643bfc93b3a8a8274d8010a39d8804/regex-2019.11.1-cp36-cp36m-manylinux1_x86_64.whl (643kB)\n","\u001b[K     |████████████████████████████████| 645kB 35.1MB/s \n","\u001b[?25hRequirement already satisfied: boto3 in /usr/local/lib/python3.6/dist-packages (from transformers) (1.10.18)\n","Requirement already satisfied: requests in /usr/local/lib/python3.6/dist-packages (from transformers) (2.21.0)\n","Collecting sacremoses\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/1f/8e/ed5364a06a9ba720fddd9820155cc57300d28f5f43a6fd7b7e817177e642/sacremoses-0.0.35.tar.gz (859kB)\n","\u001b[K     |████████████████████████████████| 860kB 35.7MB/s \n","\u001b[?25hCollecting sentencepiece\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/14/3d/efb655a670b98f62ec32d66954e1109f403db4d937c50d779a75b9763a29/sentencepiece-0.1.83-cp36-cp36m-manylinux1_x86_64.whl (1.0MB)\n","\u001b[K     |████████████████████████████████| 1.0MB 35.7MB/s \n","\u001b[?25hRequirement already satisfied: tqdm in /usr/local/lib/python3.6/dist-packages (from transformers) (4.28.1)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (from transformers) (1.17.4)\n","Requirement already satisfied: botocore<1.14.0,>=1.13.18 in /usr/local/lib/python3.6/dist-packages (from boto3->transformers) (1.13.18)\n","Requirement already satisfied: jmespath<1.0.0,>=0.7.1 in /usr/local/lib/python3.6/dist-packages (from boto3->transformers) (0.9.4)\n","Requirement already satisfied: s3transfer<0.3.0,>=0.2.0 in /usr/local/lib/python3.6/dist-packages (from boto3->transformers) (0.2.1)\n","Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests->transformers) (3.0.4)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests->transformers) (2019.9.11)\n","Requirement already satisfied: idna<2.9,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests->transformers) (2.8)\n","Requirement already satisfied: urllib3<1.25,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests->transformers) (1.24.3)\n","Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from sacremoses->transformers) (1.12.0)\n","Requirement already satisfied: click in /usr/local/lib/python3.6/dist-packages (from sacremoses->transformers) (7.0)\n","Requirement already satisfied: joblib in /usr/local/lib/python3.6/dist-packages (from sacremoses->transformers) (0.14.0)\n","Requirement already satisfied: docutils<0.16,>=0.10 in /usr/local/lib/python3.6/dist-packages (from botocore<1.14.0,>=1.13.18->boto3->transformers) (0.15.2)\n","Requirement already satisfied: python-dateutil<2.8.1,>=2.1; python_version >= \"2.7\" in /usr/local/lib/python3.6/dist-packages (from botocore<1.14.0,>=1.13.18->boto3->transformers) (2.6.1)\n","Building wheels for collected packages: sacremoses\n","  Building wheel for sacremoses (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for sacremoses: filename=sacremoses-0.0.35-cp36-none-any.whl size=883999 sha256=7142151a45c8a7d2e25d6b4c11f6cf74cef9d93dc9d657e3a87675ab9835300c\n","  Stored in directory: /root/.cache/pip/wheels/63/2a/db/63e2909042c634ef551d0d9ac825b2b0b32dede4a6d87ddc94\n","Successfully built sacremoses\n","Installing collected packages: regex, sacremoses, sentencepiece, transformers\n","Successfully installed regex-2019.11.1 sacremoses-0.0.35 sentencepiece-0.1.83 transformers-2.1.1\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"74Zhe5YtnQnl","colab_type":"code","outputId":"93c574d9-9eb2-4392-8a84-bec569180a74","executionInfo":{"status":"ok","timestamp":1574786447493,"user_tz":300,"elapsed":13538,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}},"colab":{"base_uri":"https://localhost:8080/","height":63}},"source":["import os\n","import gzip\n","from Bio import SeqIO\n","from Bio.Data import CodonTable\n","\n","import torch\n","\n","from transformers import GPT2Config, GPT2Model, GPT2LMHeadModel"],"execution_count":2,"outputs":[{"output_type":"display_data","data":{"text/html":["<p style=\"color: red;\">\n","The default version of TensorFlow in Colab will soon switch to TensorFlow 2.x.<br>\n","We recommend you <a href=\"https://www.tensorflow.org/guide/migrate\" target=\"_blank\">upgrade</a> now \n","or ensure your notebook will continue to use TensorFlow 1.x via the <code>%tensorflow_version 1.x</code> magic:\n","<a href=\"https://colab.research.google.com/notebooks/tensorflow_version.ipynb\" target=\"_blank\">more info</a>.</p>\n"],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"ybzxn8OIl1h9","colab_type":"code","outputId":"0c0afe57-10da-4975-cbee-0fe32ec67840","executionInfo":{"status":"ok","timestamp":1574786465139,"user_tz":300,"elapsed":21129,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}},"colab":{"base_uri":"https://localhost:8080/","height":122}},"source":["# Load the Drive helper and mount\n","from google.colab import drive\n","\n","# This will prompt for authorization.\n","drive.mount('/content/drive')"],"execution_count":3,"outputs":[{"output_type":"stream","text":["Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly\n","\n","Enter your authorization code:\n","··········\n","Mounted at /content/drive\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"Rxw8EzEgrH9x","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"JSdSPKL1mJnJ","colab_type":"code","colab":{}},"source":["def check_known_protein(protein_description):\n","    \"\"\"filter out hypothetical and putative proteins\"\"\"\n","    is_hypothetical = ('hypoth' in protein_description or 'hetical' in protein_description) # deals with most common misspellings of \"hypothetical\"\n","    is_putative = 'putative' in protein_description\n","    is_known = (not is_hypothetical) and (not is_putative)\n","    return is_known\n","\n","def get_CDS_locations(genome_path):\n","    \"\"\"exctracts gene locations and seqs from known protein coding genes, only non-hypothetical\"\"\"\n","    gene_loc_list = []\n","    gene_seq_list = []\n","    with gzip.open(genome_path, 'rt') as handle: # open file\n","        for record in SeqIO.parse(handle, \"genbank\"): # parse each record within file\n","            feature_list = record.features\n","            CDS_list = [x for x in feature_list if x.type=='CDS'] # get all CDS features in record\n","            for CDS in CDS_list:\n","                try:\n","                    protein_description = CDS.qualifiers['product'][0]\n","                    if check_known_protein(protein_description): # only use non-hypothetical proteins\n","                        gene_loc = CDS.location\n","                        gene_loc_list.append(gene_loc)\n","\n","                        gene_seq = CDS.extract(record.seq) # extract locations of CDS\n","                        gene_seq_list.append(gene_seq)\n","                except:\n","                    pass\n","\n","    return gene_loc_list, gene_seq_list"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"uSkSY75bmJjS","colab_type":"code","outputId":"dcacb6a4-bfe3-48e7-c4cf-ecf43fae10b9","executionInfo":{"status":"ok","timestamp":1574787746394,"user_tz":300,"elapsed":2461,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}},"colab":{"base_uri":"https://localhost:8080/","height":51}},"source":["# load fasta sequence\n","d = \"drive/My Drive/Colab Notebooks/smaug/data/ecoli_MG1655\"\n","genome_path = os.path.join(d, \"GCF_000005845.2_ASM584v2_genomic.fna.gz\")\n","\n","with gzip.open(genome_path, 'rt') as handle:\n","    for record in SeqIO.parse(handle, \"fasta\"):\n","        seq = record.seq\n","\n","# get gene locations\n","gbff_path = os.path.join(d, \"GCF_000005845.2_ASM584v2_genomic.gbff.gz\")\n","gene_loc_list, gene_seq_list = get_CDS_locations(gbff_path)\n","\n","# total genes and example loc\n","print(len(gene_loc_list))\n","print(gene_loc_list[300])"],"execution_count":7,"outputs":[{"output_type":"stream","text":["3532\n","[381350:381716](+)\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"I_WPvmSIC9ZA","colab_type":"code","outputId":"f7d2abdc-838f-48d5-d5c4-a31d6a605875","executionInfo":{"status":"ok","timestamp":1574786487027,"user_tz":300,"elapsed":8213,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}},"colab":{"base_uri":"https://localhost:8080/","height":34}},"source":["# all code needed to produce score for \"geneness\" of amino acid sequence\n","\n","# mount\n","from google.colab import drive\n","drive.mount('/content/drive')\n","# modeldir = \"drive/My Drive/Colab Notebooks/smaug/data/models\"\n","modeldir = \"drive/My Drive/Colab Notebooks/smaug/data/models/10-25-19_morning_20\"\n","\n","# load model\n","model = GPT2LMHeadModel.from_pretrained(modeldir).to('cuda')\n","\n","# score sequence\n","def score_seq(aa_seq_string):\n","    aa_seq_string = aa_seq_string[:40] ################### TODO implement shingles\n","    \n","    token = tokenize_aa_seq(aa_seq_string)\n","    input_ids = token.unsqueeze(0).to('cuda')\n","\n","    logits = model(input_ids)[0]\n","    soft = torch.exp(logits)\n","    softsum = torch.sum(soft, axis=2).view(len(soft),-1,1)\n","    softproportion = soft/softsum\n","\n","    predval = torch.roll(torch.argmax(softproportion, axis=2), 1) # remember prediction offset by 1\n","    \n","    correctmask = predval == input_ids[:] # why no offset here?\n","    hardcorrect = torch.sum(correctmask).item()\n","    totalcorrect = logits.shape[0]*logits.shape[1]\n","\n","    softcorrect = 0\n","    for i in range(input_ids.shape[0]):\n","        for j in range(input_ids.shape[1] - 1): # remember prediction offset by 1\n","            softcorrect += softproportion[i, j, input_ids[i,j+1]]\n","    softcorrect = softcorrect.item()\n","\n","    softproportion = softcorrect/totalcorrect \n","    hardproportion = hardcorrect/totalcorrect\n","\n","    return softproportion # hard proportion might work better on long sequences\n","\n","def tokenize_aa_seq(aa_seq):\n","    \"\"\"Convert amino acid letters to integers. Can also use murphy's reduced aa alphabet later\"\"\"\n","    table = {\"L\":1,\n","             \"V\":2,\n","             \"I\":3,\n","             \"M\":4,\n","             \"C\":5,\n","             \"A\":6,\n","             \"G\":7,\n","             \"S\":8,\n","             \"T\":9,\n","             \"P\":10,\n","             \"F\":11,\n","             \"Y\":12,\n","             \"W\":13,\n","             \"E\":14,\n","             \"D\":15,\n","             \"N\":16,\n","             \"Q\":17,\n","             \"K\":18,\n","             \"R\":19,\n","             \"H\":20,\n","             \"X\":0, # get rid of these\n","             \"B\":0,\n","             \"*\":0}\n","    tokenized = torch.tensor([table[aa] for aa in aa_seq])\n","    return tokenized"],"execution_count":6,"outputs":[{"output_type":"stream","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"p6QyaQl0mJgz","colab_type":"code","colab":{}},"source":["# find all ORFs in all reading frames\n","def find_ORFs(aa_seq, minimum_length):\n","    \"\"\"find positions of all open reading frames in all reading frames\"\"\"\n","    # ASSUMES AMINO ACID SEQUENCE IS 5' to 3'\n","    ORF_startstop = []\n","    started = False\n","    startpos = -1\n","    for i, c in enumerate(aa_seq):\n","        if c == \"M\" and not started:\n","            startpos = i\n","            started = True\n","            continue\n","        if c == \"*\":\n","            if started and (i-startpos >= minimum_length):\n","                ORF_startstop.append((startpos, i))\n","            started = False\n","    return ORF_startstop # TODO may be better to use dict rather than tuple list\n","\n","\n","# amino acid seqs for all 6 reading frames\n","# trimming last nucs to ensure multiple of 3\n","aa_seq_0f = str(seq[:-(len(seq)%3)].translate(table=11, to_stop=False))\n","aa_seq_1f = str(seq[1:-((len(seq)-1)%3)].translate(table=11, to_stop=False))\n","aa_seq_2f = str(seq[2:-((len(seq)-2)%3)].translate(table=11, to_stop=False))\n","aa_seq_0r = str(seq.reverse_complement()[:-(len(seq)%3)].translate(table=11, to_stop=False)) #TODO cut end off before or after reverse complement?\n","aa_seq_1r = str(seq[1:].reverse_complement()[:-((len(seq)-1)%3)].translate(table=11, to_stop=False))\n","aa_seq_2r = str(seq[2:].reverse_complement()[:-((len(seq)-2)%3)].translate(table=11, to_stop=False))\n","\n","# trim trailing aa's to get multiple of 3\n","aa_seq_0f = aa_seq_0f[:-(len(aa_seq_0f)%3)]\n","aa_seq_1f = aa_seq_0f[:-(len(aa_seq_1f)%3)]\n","aa_seq_2f = aa_seq_0f[:-(len(aa_seq_2f)%3)]\n","aa_seq_0r = aa_seq_0f[:-(len(aa_seq_0r)%3)]\n","aa_seq_1r = aa_seq_0f[:-(len(aa_seq_1r)%3)]\n","aa_seq_2r = aa_seq_0f[:-(len(aa_seq_2r)%3)]\n","\n","\n","# ORFs starts and stops in all reading frames, use to access sequence\n","minimum_length = 10\n","ORF_0f = find_ORFs(aa_seq_0f, minimum_length)\n","ORF_1f = find_ORFs(aa_seq_1f, minimum_length)\n","ORF_2f = find_ORFs(aa_seq_2f, minimum_length)\n","ORF_0r = find_ORFs(aa_seq_0r, minimum_length)\n","ORF_1r = find_ORFs(aa_seq_1r, minimum_length)\n","ORF_2r = find_ORFs(aa_seq_2r, minimum_length)\n","\n","# standardize coordinates of all ORFs, use to check for overlaps\n","l = len(ORF_0f) ### TODO aa overlap vs nucleotide overlap is complicated, think about how to make this exact and formal, though this may not matter toooo much in real data\n","ORF_0f_standard = ORF_0f\n","ORF_1f_standard = [(x[0]+1, x[1]+1) for x in ORF_1f]\n","ORF_2f_standard = [(x[0]+2, x[1]+2) for x in ORF_2f]\n","ORF_0r_standard = [(l-x[1]-1, l-x[0]-1) for x in ORF_0r] # NOTE first number is actually stop codon in all reverse frames\n","ORF_1r_standard = [(l-x[1]-1-1, l-x[0]-1-1) for x in ORF_1r]\n","ORF_2r_standard = [(l-x[1]-1-2, l-x[0]-1-2) for x in ORF_2r]\n","\n","# # example seq access\n","# foo = int(ORF_2f[0][0])\n","# bar = int(ORF_2f[0][1])\n","# print(foo, bar)\n","# print(aa_seq_2f[foo:bar+1])\n"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"a5tIldcubqc-","colab_type":"code","colab":{}},"source":["# check if predicted gene matches 3' end, also if 5' (correct start codon is harder)\n","allstops_real = \n","allstarts_real = "],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"ZyPeTgAybqRq","colab_type":"code","outputId":"903e2dcf-d88f-44dc-aa82-bf9077b27b0b","executionInfo":{"status":"ok","timestamp":1573509486165,"user_tz":300,"elapsed":227,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}},"colab":{"base_uri":"https://localhost:8080/","height":357}},"source":["for i in range(10):\n","    print(gene_loc_list[i])\n","    print(gene_loc_list[i].start)\n"],"execution_count":0,"outputs":[{"output_type":"stream","text":["[189:255](+)\n","189\n","[336:2799](+)\n","336\n","[2800:3733](+)\n","2800\n","[3733:5020](+)\n","3733\n","[5233:5530](+)\n","5233\n","[5682:6459](-)\n","5682\n","[8237:9191](+)\n","8237\n","[9305:9893](+)\n","9305\n","[9927:10494](-)\n","9927\n","[10829:11315](+)\n","10829\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"opui67sibpAP","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"gkVS862PXnDF","colab_type":"code","colab":{}},"source":["# example scores\n","import matplotlib.pyplot as plt\n","# score_0f = [score_seq(aa_seq_0f[x[0]:x[1]][::-1]) for x in ORF_0f[:1000]]\n","# score_1f = [score_seq(aa_seq_1f[x[0]:x[1]][::-1]) for x in ORF_1f[:1000]]\n","score_0r = [score_seq(aa_seq_0r[x[0]:x[1]]) for x in ORF_0f[:300]] # TODO CHECK WHETHER REVERSE OF STRING SHOULD BE USED FOR R seqs\n"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"H6ZqjCzEXnAI","colab_type":"code","outputId":"b2354ac2-af08-464c-e14e-a4c14d687189","executionInfo":{"status":"ok","timestamp":1573509617035,"user_tz":300,"elapsed":3881,"user":{"displayName":"Markus Sommer","photoUrl":"","userId":"08083804319459268544"}},"colab":{"base_uri":"https://localhost:8080/","height":265}},"source":["# plt.hist(score_0f, bins=100)\n","# plt.show()\n","# plt.hist(score_1f, bins=100)\n","# plt.show()\n","plt.hist(score_0r, bins=20)\n","plt.show()"],"execution_count":0,"outputs":[{"output_type":"display_data","data":{"image/png":"iVBORw0KGgoAAAANSUhEUgAAAXAAAAD4CAYAAAD1jb0+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0\ndHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAPT0lEQVR4nO3dbZCddX3G8e8lgaLYSoA1k4Jx44gP\n2ArqCjpaxxofqKhJpwwFrZOxzGQ6tR1tayXaN9rpC3Q6tb7o1GZE3c5Ygaa1MNJiY9RqW4smCkiI\nFoihkgYSEaqIlQZ/fXHuNOtmw56ch835N9/PzM65H/dcc7J75b/3ff67qSokSe153LEOIEkajAUu\nSY2ywCWpURa4JDXKApekRi1byic744wzanp6eimfUpKat3379u9U1dT87Uta4NPT02zbtm0pn1KS\nmpfk7oW2ewlFkhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RGWeCS1CgLXJIataQzMbV0pjfe\nMPC5u6+8aIRJJI2LI3BJapQFLkmN6qvAk5yaZHOSbyTZmeTFSU5LsiXJHd3j8nGHlSQd0u8I/IPA\njVX1LOBcYCewEdhaVWcDW7t1SdISWbTAkzwJeBlwFUBVPVJVDwJrgdnusFlg3bhCSpIO188IfDWw\nH/hokq8l+XCSU4AVVbW3O+ZeYMVCJyfZkGRbkm379+8fTWpJUl8Fvgx4PvDnVfU84AfMu1xSVQXU\nQidX1aaqmqmqmampw/6ghCRpQP0U+D3APVV1U7e+mV6h35dkJUD3uG88ESVJC1m0wKvqXuDbSZ7Z\nbVoD3A5cD6zvtq0HrhtLQknSgvqdifnbwMeTnATsAt5Cr/yvTXI5cDdwyXgiSpIW0leBV9XNwMwC\nu9aMNo4kqV/OxJSkRlngktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtc\nkhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RGWeCS1CgLXJIaZYFLUqMscElqlAUuSY2ywCWp\nURa4JDXKApekRi3r56Aku4HvA48CB6pqJslpwDXANLAbuKSqHhhPTEnSfEczAv/Fqjqvqma69Y3A\n1qo6G9jarUuSlsgwl1DWArPd8iywbvg4kqR+9XUJBSjgH5MU8BdVtQlYUVV7u/33AisWOjHJBmAD\nwKpVq4aMe3yZ3njDsY4gaYL1W+Avrao9SZ4MbEnyjbk7q6q6cj9MV/abAGZmZhY8RpJ09Pq6hFJV\ne7rHfcAngfOB+5KsBOge940rpCTpcIsWeJJTkvz0wWXg1cBtwPXA+u6w9cB14wopSTpcP5dQVgCf\nTHLw+L+qqhuTfAW4NsnlwN3AJeOLKUmab9ECr6pdwLkLbL8fWDOOUJKkxTkTU5IaZYFLUqMscElq\nlAUuSY2ywCWpURa4JDXKApekRlngktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ\n4JLUKAtckhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RGWeCS1Khl/R6Y5ARgG7Cnql6XZDVw\nNXA6sB14c1U9Mp6Y7ZreeMOxjiDp/6mjGYG/Ddg5Z/19wAeq6unAA8DlowwmSXpsfRV4krOAi4AP\nd+sBXgFs7g6ZBdaNI6AkaWH9jsD/FHgn8ONu/XTgwao60K3fA5y50IlJNiTZlmTb/v37hworSTpk\n0QJP8jpgX1VtH+QJqmpTVc1U1czU1NQgn0KStIB+bmK+BHhDktcCJwM/A3wQODXJsm4UfhawZ3wx\nJUnzLToCr6p3VdVZVTUNXAp8tqreBHwOuLg7bD1w3dhSSpIOM8z7wK8AfjfJnfSuiV81mkiSpH70\n/T5wgKr6PPD5bnkXcP7oI0mS+uFMTElq1FGNwHV8GHb26O4rLxpREkmPxRG4JDXKApekRlngktQo\nC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtckhrlL7PSyA3zy7D8RVhS/xyB\nS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtckhplgUtSoxYt8CQnJ/lykluS7Ejy3m77\n6iQ3JbkzyTVJThp/XEnSQf2MwH8EvKKqzgXOAy5M8iLgfcAHqurpwAPA5eOLKUmab9ECr56HutUT\nu48CXgFs7rbPAuvGklCStKC+roEnOSHJzcA+YAtwF/BgVR3oDrkHOHM8ESVJC+mrwKvq0ao6DzgL\nOB94Vr9PkGRDkm1Jtu3fv3/AmJKk+Y7qXShV9SDwOeDFwKlJDv462rOAPUc4Z1NVzVTVzNTU1FBh\nJUmH9PMulKkkp3bLjwdeBeykV+QXd4etB64bV0hJ0uH6+YMOK4HZJCfQK/xrq+pTSW4Hrk7yR8DX\ngKvGmFOSNM+iBV5VtwLPW2D7LnrXwyVJx4AzMSWpURa4JDXKApekRlngktQoC1ySGmWBS1KjLHBJ\napQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtckhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RG\nWeCS1CgLXJIatehfpT/eTW+84VhHkKQFOQKXpEZZ4JLUKAtckhq1aIEneUqSzyW5PcmOJG/rtp+W\nZEuSO7rH5eOPK0k6qJ8R+AHg96rqHOBFwFuTnANsBLZW1dnA1m5dkrREFi3wqtpbVV/tlr8P7ATO\nBNYCs91hs8C6cYWUJB3uqN5GmGQaeB5wE7CiqvZ2u+4FVhzhnA3ABoBVq1YNmlPHiWHetrn7yotG\nmESafH3fxEzyROBvgLdX1ffm7quqAmqh86pqU1XNVNXM1NTUUGElSYf0VeBJTqRX3h+vqr/tNt+X\nZGW3fyWwbzwRJUkL6eddKAGuAnZW1Z/M2XU9sL5bXg9cN/p4kqQj6eca+EuANwNfT3Jzt+3dwJXA\ntUkuB+4GLhlPREnSQhYt8Kr6ZyBH2L1mtHEkSf1yJqYkNcoCl6RGWeCS1CgLXJIaZYFLUqMscElq\nlAUuSY2ywCWpURa4JDXKApekRlngktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ\n4JLUKAtckhplgUtSoyxwSWqUBS5JjbLAJalRixZ4ko8k2ZfktjnbTkuyJckd3ePy8caUJM3Xzwj8\nY8CF87ZtBLZW1dnA1m5dkrSEFi3wqvoC8N15m9cCs93yLLBuxLkkSYtYNuB5K6pqb7d8L7DiSAcm\n2QBsAFi1atWATyctbnrjDQOfu/vKi0aYRFoaQ9/ErKoC6jH2b6qqmaqamZqaGvbpJEmdQQv8viQr\nAbrHfaOLJEnqx6AFfj2wvlteD1w3mjiSpH718zbCTwBfAp6Z5J4klwNXAq9Kcgfwym5dkrSEFr2J\nWVWXHWHXmhFnkSQdBWdiSlKjLHBJapQFLkmNssAlqVGDzsRsyjAz9HR8cBanWuQIXJIaZYFLUqMs\ncElqlAUuSY2ywCWpURa4JDXKApekRlngktQoC1ySGnVczMSUxmnYmb7O5NSgHIFLUqMscElqlAUu\nSY1q5hq4v1FQkn6SI3BJapQFLkmNssAlqVEWuCQ1ygKXpEY18y4USaPl3wE9OpP4eg01Ak9yYZJv\nJrkzycZRhZIkLW7gAk9yAvBnwC8B5wCXJTlnVMEkSY9tmBH4+cCdVbWrqh4BrgbWjiaWJGkxw1wD\nPxP49pz1e4AL5h+UZAOwoVt9KMk35+w+A/jOEBnGyWyDMdtRyvuACc3WOSxbl3kSNPG6jeD1eupC\nG8d+E7OqNgGbFtqXZFtVzYw7wyDMNhizDcZsgznesw1zCWUP8JQ562d12yRJS2CYAv8KcHaS1UlO\nAi4Frh9NLEnSYga+hFJVB5L8FvBp4ATgI1W14yg/zYKXViaE2QZjtsGYbTDHdbZU1bifQ5I0Bk6l\nl6RGWeCS1KixFPhiU+yT/FSSa7r9NyWZ7rafn+Tm7uOWJL88Sfnm7F+V5KEk75iUbEmmk/xwzuv3\noUnJ1u17bpIvJdmR5OtJTp6EbEneNOc1uznJj5OcNyHZTkwy271eO5O8a5S5hsx2UpKPdtluSfLy\nY5DtZUm+muRAkovn7Vuf5I7uY/2EZbsxyYNJPjV0kKoa6Qe9G5p3AU8DTgJuAc6Zd8xvAh/qli8F\nrumWnwAs65ZXAvsOrk9Cvjn7NwN/DbxjUrIB08Bto/73HFG2ZcCtwLnd+unACZOQbd4xPw/cNUGv\n2xuBq7vlJwC7gekJyfZW4KPd8pOB7cDjljjbNPBc4C+Bi+dsPw3Y1T0u75aXT0K2bt8a4PXAp4bN\nMo4ReD9T7NcCs93yZmBNklTVw1V1oNt+MjCOO6wD5wNIsg74FnC077gZe7YxGybbq4Fbq+oWgKq6\nv6oenZBsc13WnTtKw2Qr4JQky4DHA48A35uQbOcAnwWoqn3Ag8AoJ60smq2qdlfVrcCP5537GmBL\nVX23qh4AtgAXTkg2qmor8P1RBBlHgS80xf7MIx3TFfZ/0RuVkeSCJDuArwO/MafQj3m+JE8ErgDe\nO+JMQ2fr9q1O8rUk/5TkFyYo2zOASvLp7sfKd05Qtrl+FfjEBGXbDPwA2Av8B/DHVfXdCcl2C/CG\nJMuSrAZewE9O7FuKbOM4dxI+f98m7veBV9VNwHOSPBuYTfIPVfXfxzpX5z3AB6rqoaUZ9B6VvcCq\nqro/yQuAv0vynKoa5YhtUMuAlwIvBB4GtibZ3o1EJkKSC4CHq+q2Y51ljvOBR4GfpXcp4ItJPlNV\nu45tLAA+Ajwb2AbcDfwrvaxaQuMYgfczxf7/jul+PHwScP/cA6pqJ/AQ8HMTlO8C4P1JdgNvB96d\n3mSmY56tqn5UVfcDVNV2etfonjEJ2eiNUL5QVd+pqoeBvweePyHZDrqU0Y++h832RuDGqvqf7jLF\nvzDayxTDfL0dqKrfqarzqmotcCrw70ucbRznTsLn79+oLuzPuUC/jN5Ng9UcusD/nHnHvJWfvDFy\nbbe8mkM3MZ8K/CdwxqTkm3fMexj9TcxhXrspuhuD9G6u7AFOm5Bsy4Gv0t2kBj4DXDQJ2br1x3Wv\n19Mm7PvhCg7dKDwFuB147oRkewJwSrf8Knr/QS/p6zbn2I9x+E3Mb3Vfd8u75SX9XjhStjnbX84I\nbmKO9It1TrjX0vvf+C7gD7ptfwi8oVs+md67OO4EvnzwGwd4M72bgzd33/DrJinfvM/xHkZc4EO+\ndr8y77V7/aRk6/b9WpfvNuD9E5bt5cC/jeNrbch/0yd223fQK+/fn6Bs08A3gZ30/kN+6jHI9kJ6\nP939gN5PLDvmnPvrXeY7gbdMWLYvAvuBH3bHvGbQHE6ll6RGORNTkhplgUtSoyxwSWqUBS5JjbLA\nJalRFrgkNcoCl6RG/S/a40bji6iwYwAAAABJRU5ErkJggg==\n","text/plain":["<Figure size 432x288 with 1 Axes>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"u5s-xsZstbls","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"fI36qtZotbhL","colab_type":"code","colab":{}},"source":["# greedy search to find non-overlapping genes based solely on length\n"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"o0BGMhxOtbfb","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"QgjEFYestbeS","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"38wZrCrbtba1","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"n_6Qu066tbZI","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"GhqLi4NPtbW5","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"kjUelzoxtbVc","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"YuImHoPitbSV","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"BYJ9De9_tbQ5","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"lVzYDE45tbNn","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"w1m8FefnmJef","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]}]}